node/core/chain-selection/src/backend.rs

```diff
@@ -0,0 +1,238 @@
+//! struct which allows in-memory changes to be applied on top of a
+//! [`Backend`], maintaining consistency between queries and temporary writes,
+//! before any commit to the underlying storage is made.
+
+use polkadot_primitives::v1::{BlockNumber, Hash, Header, ConsensusLog};
+
+use std::collections::HashMap;
+
+use crate::{Error, LeafEntrySet, BlockEntry, Timestamp};
+
+pub(super) enum BackendWriteOp {
+	WriteBlockEntry(BlockEntry),
+	WriteBlocksByNumber(BlockNumber, Vec<Hash>),
+	WriteViableLeaves(LeafEntrySet),
+	WriteStagnantAt(Timestamp, Vec<Hash>),
+	DeleteBlocksByNumber(BlockNumber),
+	DeleteBlockEntry(Hash),
+	DeleteStagnantAt(Timestamp),
+}
+
+/// An abstraction over backend storage for the logic of this subsystem.
+pub(super) trait Backend {
+	/// Load a block entry from the DB.
+	fn load_block_entry(&self, hash: &Hash) -> Result<Option<BlockEntry>, Error>;
+	/// Load the active-leaves set.
+	fn load_leaves(&self) -> Result<LeafEntrySet, Error>;
+	/// Load the stagnant list at the given timestamp.
+	fn load_stagnant_at(&self, timestamp: Timestamp) -> Result<Vec<Hash>, Error>;
+	/// Load all stagnant lists up to and including the given unix timestamp
+	/// in ascending order.
+	fn load_stagnant_at_up_to(&self, up_to: Timestamp)
+		-> Result<Vec<(Timestamp, Vec<Hash>)>, Error>;
+	/// Load the earliest kept block number.
+	fn load_first_block_number(&self) -> Result<Option<BlockNumber>, Error>;
+	/// Load blocks by number.
+	fn load_blocks_by_number(&self, number: BlockNumber) -> Result<Vec<Hash>, Error>;
+
+	/// Atomically write the list of operations, with later operations taking precedence over prior.
+	fn write<I>(&mut self, ops: I) -> Result<(), Error>
+		where I: IntoIterator<Item = BackendWriteOp>;
+}
+
+/// An in-memory overlay over the backend.
+///
+/// This maintains read-only access to the underlying backend, but can be
+/// converted into a set of write operations which will, when written to
+/// the underlying backend, give the same view as the state of the overlay.
+pub(super) struct OverlayedBackend<'a, B: 'a> {
+	inner: &'a B,
+
+	// `None` means 'deleted', missing means query inner.
+	block_entries: HashMap<Hash, Option<BlockEntry>>,
+	// `None` means 'deleted', missing means query inner.
+	blocks_by_number: HashMap<BlockNumber, Option<Vec<Hash>>>,
+	// 'None' means 'deleted', missing means query inner.
+	stagnant_at: HashMap<Timestamp, Option<Vec<Hash>>>,
+	// 'None' means query inner.
+	leaves: Option<LeafEntrySet>,
+}
+
+impl<'a, B: 'a + Backend> OverlayedBackend<'a, B> {
+	pub(super) fn new(backend: &'a B) -> Self {
+		OverlayedBackend {
+			inner: backend,
+			block_entries: HashMap::new(),
+			blocks_by_number: HashMap::new(),
+			stagnant_at: HashMap::new(),
+			leaves: None,
+		}
+	}
+
+	pub(super) fn load_block_entry(&self, hash: &Hash) -> Result<Option<BlockEntry>, Error> {
+		if let Some(val) = self.block_entries.get(&hash) {
+			return Ok(val.clone())
+		}
+
+		self.inner.load_block_entry(hash)
```

Comment:
I assume the underlying backend has some caching, which rocksdb does, so this is OK.

Try implementing `fn propagate_viability_update` without the overlay if you don't think it's necessary!

Reply:
If we'd like to cache on load, and find that rocksdb cache is not sufficient, I think that the best approach to take would be to use trait composability.

So we'd write a 

```rust
struct CachedBackend<B> { cache, inner: B }

impl<B: Backend> Backend for CachedBackend<B> { ... }
```

This would allow us to implement a targeted cache without affecting the logic of the overlay.

Reply:
I actually initially tried to implement `Backend for OverlayedBackend<B: Backend>` but the `first_block_number` and `stagnant_at_up_to` methods proved pretty much impossible.

Reply:
Ok, fair enough. I don't think it's worth blocking this PR on this concern, but it's worth revisiting once we merge all the dbs.